llm/kv_cache_test.go

```diff
@@ -0,0 +1,188 @@
+type testGGML struct {
+	kv KV
+}
+
+func (g *testGGML) KV() KV {
+	return g.kv
+}
+
+func TestKVCacheQuantization(t *testing.T) {
+	tests := []struct {
+		name         string
+		kvData       map[string]any
+		gpus         discover.GpuInfoList
+		envKVCache   string // simulated environment variable value
+		envFlashAttn string // simulated environment variable value
+		expected     string
+	}{
+		{
+			name: "empty cache type",
+			kvData: map[string]any{
+				"general.architecture": "llama",
+			},
+			gpus: discover.GpuInfoList{
+				{Library: "cuda", DriverMajor: 8},
+			},
+			envKVCache:   "",
+			envFlashAttn: "false",
+			expected:     "f16",
+		},
+		{
+			name: "invalid cache type",
+			kvData: map[string]any{
+				"general.architecture": "llama",
+			},
+			gpus: discover.GpuInfoList{
+				{Library: "cuda", DriverMajor: 8},
+			},
+			envKVCache:   "invalid",
+			envFlashAttn: "true",
+			expected:     "f16",
+		},
+		{
+			name: "embedding model with q4_0",
+			kvData: map[string]any{
+				"general.architecture": "bert",
+				"bert.pooling_type":    "mean",
+			},
+			gpus: discover.GpuInfoList{
+				{Library: "cuda", DriverMajor: 8},
+			},
+			envKVCache:   "q4_0",
+			envFlashAttn: "true",
+			expected:     "f16",
+		},
+		{
+			name: "embedding model with f32",
+			kvData: map[string]any{
+				"general.architecture": "bert",
+				"bert.pooling_type":    "mean",
+			},
+			gpus: discover.GpuInfoList{
+				{Library: "cuda", DriverMajor: 8},
+			},
+			envKVCache:   "f32",
+			envFlashAttn: "true",
+			expected:     "f32",
+		},
+		{
+			name: "non-embedding model with q4_0 and flash attention",
+			kvData: map[string]any{
+				"general.architecture":         "llama",
+				"llama.attention.key_length":   uint32(32),
+				"llama.attention.value_length": uint32(32),
+			},
+			gpus: discover.GpuInfoList{
+				{Library: "cuda", DriverMajor: 8},
+			},
+			envKVCache:   "q4_0",
+			envFlashAttn: "true",
+			expected:     "q4_0",
+		},
+		{
+			name: "non-embedding model with q4_0 without flash attention",
+			kvData: map[string]any{
+				"general.architecture": "llama",
+			},
+			gpus: discover.GpuInfoList{
+				{Library: "cuda", DriverMajor: 6}, // older driver that doesn't support flash attention
+			},
+			envKVCache:   "q4_0",
+			envFlashAttn: "true",
+			expected:     "f16",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Set up environment for the test
+			t.Setenv("OLLAMA_KV_CACHE_TYPE", tt.envKVCache)
+			t.Setenv("OLLAMA_FLASH_ATTENTION", tt.envFlashAttn)
+
+			ggml := &testGGML{kv: tt.kvData}
+			result := KVCacheQuantization(ggml, tt.gpus)
+			if result != tt.expected {
+				t.Errorf("expected %s, got %s", tt.expected, result)
+			}
+		})
+	}
+}
+
+// ValidateFlashAttentionSupport checks if flash attention is supported by the model and hardware
+func ValidateFlashAttentionSupport(ggml GGMLModel, gpus discover.GpuInfoList, flashAttnRequested bool) bool {
```

Comment:
I think this is unused (minus tests :)?

Reply:
Ah yes, a remnant from before refactoring, cleaned up and refactored the test to perform the logic directly.